### PR TITLE
Fix NURBS boolean failures from unhandled edge-on-edge classification (fixes #1619, part of #1091)

### LINKED ISSUE #683: Invalid "The mesh has naked edges" error
### System information

SolveSpace version: 3.0~6894b0c9

Operating system: Windows 10

### Expected behavior

I should not see any errors

### Actual behavior
After extruding
image

After save - load
image

If remove last extruding layer  and create extruding again the error will be shown.
image

### Additional information
[wrong.zip](https://github.com/solvespace/solvespace/files/5112393/wrong.zip)

### LINKED ISSUE #728: Unexpected boolean operation fails in extrusion
### System information

SolveSpace version: master

Operating system: debian 9 amd64

### Expected behavior

Successful results

### Actual behavior

Boolean operation failure error, some red edges highlighted on the extrusion, and a spurious chord appears from intersecting geometry.

### Additional information

Open tapered-rafter-12-to-10-to-8-gusseted-truss.slvs in the attached zip to repro:

[gusseted-truss.zip](https://github.com/solvespace/solvespace/files/5315184/gusseted-truss.zip)

### LINKED ISSUE #1091: NURBS failure with countersunk holes
### System information

- **SolveSpace version:** master @ 56719415dea4de7af750227653362f4c8cad78f8 (3.0)
- **Operating system:** macOS 11.5.1 (Big Sur)

### Expected behavior

Opened [turc-nurbs-1.slvs.zip](https://github.com/solvespace/solvespace/files/6986416/turc-nurbs-1.slvs.zip)
Expected no NURBS issues

### Actual behavior

The 'board-1-inset' group shows "The Boolean operation failed", there are angry red lines around some of the geometry from this group and Analyze->Naked Edges reports "The mesh has naked edges (NOT okay, invalid). 10 problematic edges, bad." On the console I get:

```
failed: I=20, avoid=7
print 4 edges
failed to assemble polygon to trim nurbs surface in uv space
```

### Additional information

The problematic model has a couple of holes through a solid board (called "top-plate" in the model), modelled as a difference extrude ("board-1-holes" being the extrude of the "board-1" sketch). These holes are countersunk by adding a second sketch of slightly-larger holes and a second difference extrude that doesn't go all the way through the board ("board-1-inset-sketch" and "board-1-inset"). I've simplified board-1-inset-sketch to reduce the number of holes, and only countersink the top-left hole, to hopefully make the test case easier. When I tried removing some more construction geometry from "board-1" the problem went away and I'm not sure why.

The overall project that this model is part of quickly degenerated into having to tick "Force NURBS surfaces to triangle mesh" early on and is pretty complex and incredibly slow to recalculate (Generate::ALL takes 30s). I read a comment on another issue suggesting that the triangle meshing might be contributing to the performance issues, hence why I'm raising this to see if I can get the whole thing to work in NURBS.

### LINKED ISSUE #1305: Surface experiments
3.1 70bde63c Version on Manjaro.

While drawing stuff my old friend surface problem showed up. This time I documented it a little for any debug use. 

Subtractive chamfer on corner drawn from the inside (0,0) worked but did not work from outside (inf,inf).  

errors
other side
When drawing the other side also from the inside what previously worked: 

other_side from innerplane

But partial is good:

other_side_partial

This is inconsistent as the with a little shaking of the point or going in or out of extrusion the red warning disappears. This fuzzy state goes away when the chord tolerance is raised to 0.01 from 0.1 and 10 segs instead 20.
Strange edges it got there too:

weird_curve

The first example lets me think of a failed inside outside detection for surface creation?

### LINKED ISSUE #1367: Edge_Fail - simple NURBS fail
### System information

- **SolveSpace version:** All
- **Operating system:** All

### Expected behavior

The attached sketch should not have naked edges.

### Actual behavior

The sketch fails depending on the location of points P1 - P4.

### Additional information

It appears that he real issue is misclassification of the edge(s) at the origin. Failure always show the short red edge at the origin in addition to a red rectangle where surface S3 protrudes from the first group. Moving P1 to the lleft can ffix it, as can a number of other movements ot P1-P4.

For discussion Lets call Surfaces S1 - S4 the surfaces connecting the origin to P1-P4. Also Edge normals (for in -surface edge normals) EN1-EN4 would be the in-surface normals of the main edge (pointing along each surface from the origin).  Also surface normals SN1-SN4 would be the outward surface normals - these should bee perpendicular to their ENx edge normal.

[Edge_Fail.zip](https://github.com/solvespace/solvespace/files/11412489/Edge_Fail.zip)

### LINKED ISSUE #1390: "failed to assemble polygon to trim nurbs surface in uv space" with a simple wedge cut-out
### System information

- **SolveSpace version:** 3.1~70bde63c
- **Operating system:** Gentoo Linux

### Expected behavior

There is no error if extrusion does not extend all the way through the other group:

image

### Actual behavior

image

### Additional information

[wedge-diff-nurbs.slvs.gz](https://github.com/solvespace/solvespace/files/12318087/wedge-diff-nurbs.slvs.gz)

### LINKED ISSUE #1398: unstable "failed to assemble polygon to trim nurbs surface in uv space"
### System information

- * **SolveSpace version:** 3.1~70bde63c
- * **Operating system:** Gentoo Linux

### Actual behavior

...depends on extrusion depth.

nurbs-2

### Additional information

[nurbs-2.slvs.gz](https://github.com/solvespace/solvespace/files/12435894/nurbs-2.slvs.gz)

### LINKED ISSUE #1619: NURBS failure when ClassifyEdge fails (was: Difference fail depending on compilation options (was: Difference fail with OpenMP))
This was first [reported on the forum](https://solvespace.com/forum.pl?action=viewthread&parent=6129&tt=1755262509) using the attached file.

[Issue_OpenMP.zip](https://github.com/user-attachments/files/21822164/Issue_OpenMP.zip)

You can make the problem go away by:
1) Dragging the original extrusion the other direction
2) Using a version of Solvespace built without OpenMP
3) Open it using the web version (which does not use OpenMP)

Treating it as a concurrency problem I have been able to make the issue go away by commenting out the following line in SShell::CopySurfacesTrimAgainst()
```
#pragma omp parallel for
```
That function calls MakeCopyTrimAgainst() and inside of that we can "fix" things by adding #pragma omp critical  to either of the two big loops [1 here](https://github.com/solvespace/solvespace/blob/74a13ef61c490a400eaa49c5335622d91ada12fc/src/srf/boolean.cpp#L603) or the one under it [2 here](https://github.com/solvespace/solvespace/blob/74a13ef61c490a400eaa49c5335622d91ada12fc/src/srf/boolean.cpp#L637).

I don't understand how it can be fixed by making a critical section for either of two loops. At this point I don't have any idea what's causing it.

### PR DESCRIPTION
**Pull request by BoykoNeov**
# Fix NURBS boolean failures from unhandled edge-on-edge classification (fixes #1619, part of #1091)

## Summary

Fixes #1619 ("failed to assemble polygon to trim nurbs surface", initially
suspected to be an OpenMP concurrency bug), and the unhandled knife-edge
classification case found while testing against the #1091 three-prisms
model.

`SShell::ClassifyEdge()` had two failure paths in the edge-on-edge case:

1. **Tangent at a shell edge** (#1619): when the edge is tangent to the shell
   at the shell's edge, it did `return false` without writing its
   `indir`/`outdir` outputs — and both call sites in
   `SSurface::MakeCopyTrimAgainst()` ignore the return value, so `KeepEdge()`
   was called with **uninitialized stack values**. Whether the edge was kept
   depended on whatever was on the stack, which is why the outcome flipped
   with build type, OpenMP, and sanitizers (see @iscgar's debug diff in the
   issue, and @ruevs' build-mode table). @jwesthues anticipated exactly this
   in https://github.com/solvespace/solvespace/issues/1091#issuecomment-904048397:
   "I suspect some paths now access uninitialized memory for
   `{in,out}dir_shell`."

2. **Knife edge** (#1091, `NURBS_ThreePrisms.slvs`): when regions of the
   shell join along a knife edge, an edge lying on that line intersects
   *four* (in general, any even number of) coincident shell edges. The
   `edge_inters == 2` block was skipped entirely, the code fell through to
   the ray-casting fallback, and the wrong classification broke the trim
   polygon (naked edges, "failed to assemble polygon").

The attached `Issue_OpenMP.slvs` hits path 1 because the subtracted wedge is
sketched exactly on the box's face and its hypotenuse plane touches the box's
bottom-front edge tangentially.

## The fixes

**`src/srf/raycast.cpp`, tangent case** — implement the missing
classification instead of returning false. When our edge is tangent to the
shell at the shell's edge, the two faces meeting at that edge lie on the same
side of our surface, so the regions on both sides of our edge get the same
class:

- **outside** the shell if the shell's edge is convex,
- **inside** if it's concave (reflex).

The vectors already computed in this function give this directly. Since
`inter_edge_n[i]` points away from face *i*'s material, the shell's edge is
convex iff face 1's material direction lies behind face 0's plane:

```cpp
Class c = (inter_edge_n[1].Dot(inter_surf_n[0]) > 0) ?
                Class::SURF_OUTSIDE : Class::SURF_INSIDE;
*indir  = c;
*outdir = c;
```

(Orientation convention cross-checked against the function's own
coincident-face branches and against `EdgeNormalsWithinSurface()`, whose
actual math treats `n×d` as the *outer* edge normal — the "inner normal"
comment at its line 457 is stale; note `pin = PointAt(muv.Minus(enuv))` and
the commented-out `*enout = enxyz`.)

**`src/srf/raycast.cpp`, knife-edge case** — handle `edge_inters >= 4`
(even). The faces along the knife edge divide the pencil of directions
perpendicular to the edge at p into sectors that are alternately inside and
outside the shell. Classify each side of our edge by the face *angularly
closest* to it: the side is outside the shell iff it lies on the outer side
of that face, and if it lies *along* the nearest face (within `DOTP_TOL`),
it's coincident, same/opposite by the surface normals — the same answers the
`edge_inters == 2` block produces, generalized to any number of faces. (This
is jwesthues' suggested experiment from #1091 made robust: rather than
"picking any pair and classifying as if n=2", the nearest face always bounds
the sector the queried direction lies in, so no arbitrary pairing is needed;
it also tolerates duplicate matches from split edge chains.)

**`src/srf/boolean.cpp`** — defensive hardening at both call sites:
initialize `indir_shell`/`outdir_shell` to a deterministic value and print a
debug warning if `ClassifyEdge()` ever returns false again, so any future
unhandled classification case fails loudly and reproducibly instead of
depending on uninitialized memory (per jwesthues' recommendation in #1091).

## Verification

All on Windows 11, MSVC 19.44 x64, `-G Ninja`, `ENABLE_OPENMP=ON`:

1. **Reproduced first** (unfixed master, Debug): `solvespace-cli export-mesh`
   on `Issue_OpenMP.slvs` prints `failed: I=3, avoid=2` and `failed to
   assemble polygon to trim nurbs surface in uv space` — same surface 3 as in
   the issue.
2. **Mechanism proven**: initializing the two variables with a "poison" value
   (`SURF_INSIDE`) made the failure 100% deterministic in every build config,
   with `ClassifyEdge()` returning false 3 times, all on surface 3.
3. **Geometry verified against ground truth**: the expected volume computed
   analytically from the sketch data is
   (59.3·80 − ½·59.3·14.667931)·59.3 = **255,529.4 mm³**; the fixed build
   exports a mesh of exactly 255,529.39 mm³ with the expected 6-quad-face
   topology (12 triangles). An unfixed *Release* build (the config that was
   "OK" by stack-garbage luck) produces the identical triangle set.
4. **Concave branch verified too, and covered by a testsuite case**
   (`test/group/boolean_tangent_edge/`): an L-section extrusion minus a
   triangular prism whose face plane passes tangentially through the L's
   concave vertical edge. Instrumentation confirms this one model exercises
   **both** new tangent branches at the same edge — the prism's edge against
   the L classifies concave → `SURF_INSIDE`, and the L's edge against the
   prism (where that edge is a convex wedge of the prism) classifies convex →
   `SURF_OUTSIDE`. The test checks that the mesh is watertight and
   non-self-intersecting and that its volume matches the analytic value
   (296,000 mm³) from the sketch. On unfixed code with the poison-init from
   (2), the same model fails deterministically — naked edges, wrong geometry
   (264,000 mm³), and the "failed to assemble polygon" error.
5. **Knife-edge case verified and covered by a testsuite case**
   (`test/group/boolean_knife_edge/`, fixture = ruevs' `NURBS_ThreePrisms`
   model from #1091): previously 12 × `bad, edge_inters=4`, surface 13
   failed to assemble, 21 triangles with 5 naked edges. Fixed: clean export,
   watertight, 24 triangles, volume exactly the analytic
   (½·1·1.6126052 + ½·1.5498346·1)·1.0 + ½·1·1·0.4983402 = **1.8303900 mm³**
   (independently cross-checked against a force-to-triangle-mesh export of
   the same model: identical volume). Identical triangle sets across Debug,
   Release, and repeated OpenMP runs. The test fails with naked edges when
   the knife-edge classification is reverted.
6. **No regressions**: full test suite passes in Debug and Release
   (now 259 cases / 907 checks); `NURBS_TwoPrisms.slvs` still exports
   cleanly, and repeated OpenMP runs of all repro models produce identical
   geometry. (I avoided `CHECK_SAVE`/render comparisons in both new tests on
   purpose: the Boolean-derived surface cache contains a few last-ulp values
   that could differ across compilers/FP-contraction settings, and
   `PrepareSavefile` does not normalize `Curve`/`CCtrl` lines.)

## Notes / limitations

- **Stacked on the companion race-fix PR** ("Fix an OpenMP use-after-free on
  into->curve in AddExactIntersectionCurve()"): stress-testing the
  three-prisms model exposed a pre-existing OpenMP use-after-free in
  `SSurface::AddExactIntersectionCurve()` that intermittently crashes
  Booleans on models with many identical exact intersection curves — and the
  new `boolean_knife_edge` test model is the heaviest known trigger (the
  test suite segfaults in roughly half of Debug OpenMP runs without that
  fix). The race is an independent, pre-existing bug on master, so it's a
  separate PR; this branch is based on it and it should merge first.
- The other #1091 sub-issues (tangent curve-surface intersections through a
  cylinder's silhouette, the "proximity" failures, trim-curve splitting) are
  different mechanisms and remain open; this PR only fixes the edge-on-edge
  classification cases.
- jwesthues noted in #1091 that the Booleans assume non-degenerate geometry
  without regions joining along knife edges; the new `edge_inters >= 4`
  branch handles the classification half of that assumption. If some other
  stage still trips on a knife-edge model, it will now at least do so
  deterministically thanks to the boolean.cpp hardening.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/srf/raycast.cpp [hunk 1]
```diff
@@ -506,9 +506,84 @@ bool SShell::ClassifyEdge(Class *indir, Class *outdir,
...
+            // behind face 0's plane.
+            Class c = (inter_edge_n[1].Dot(inter_surf_n[0]) > 0) ?
+                            Class::SURF_OUTSIDE : Class::SURF_INSIDE;
+            *indir  = c;
+            *outdir = c;
+        }
+        return true;
+    }
+
+    if(edge_inters >= 4 && edge_inters % 2 == 0) {
```
Comment (ruevs) [inline src/srf/raycast.cpp:525]: @jwesthues @phkahler Please look at this section - I stepped through it with the debugger, I think I understand it, it definitely works; but I'm still not 100% sure it's "the right thing" to do.

I'll attach a test model (in addition to https://github.com/solvespace/solvespace/files/7030785/NURBS_ThreePrisms.zip).

File: src/srf/raycast.cpp [hunk 2]
```diff
@@ -506,9 +506,84 @@ bool SShell::ClassifyEdge(Class *indir, Class *outdir,
...
-            return false;
+            // Our edge is tangent to the shell at the shell's edge: the
+            // two faces meeting at the shell's edge lie on the same side
+            // of our surface, so the regions on both sides of our edge
+            // get the same class. That's outside the shell if the shell's
+            // edge is convex, and inside if it's concave (reflex).
+            // inter_edge_n[i] points away from face i's material, so the
+            // shell's edge is convex iff face 1's material direction lies
+            // behind face 0's plane.
+            Class c = (inter_edge_n[1].Dot(inter_surf_n[0]) > 0) ?
```
Comment (ruevs) [inline src/srf/raycast.cpp:517]: Very interesting! Was it really so simple?!

**Review by ruevs — COMMENTED** [review]
When the [LLMs generate slop it is not worth looking at it](https://github.com/solvespace/solvespace/blob/master/CONTRIBUTING.md#llm-or-ai-contributions). But this is not slop!

This is in fact _very_ good. The code is coherent, the comments are coherent, the analysis is correct. Thanks @BoykoNeov for setting [Claude Fable 5](https://platform.claude.com/docs/en/about-claude/models/introducing-claude-fable-5-and-claude-mythos-5) onto this. This is the first time when I'm truly impressed by a programming problem solved by AI.

Speaking of problem - this actually fixes two of the most commonly reported NURBS bugs (See the comments in the individual commits for details):

- [Classifying edges tangent to the shell at a shell edge](https://github.com/solvespace/solvespace/pull/1729/changes/93659278a7c637908261e71121a62189b0d94778)
- [Classifying edges lying on a knife edge of the shell, where more than two
surfaces meet edge-on-edge.](https://github.com/solvespace/solvespace/pull/1729/changes/a1422a8fae0a4daa8c11ba8b5fb0d3ebc8b24482)

For the second one I am attaching a new test model that underlines the solution:
[1729_TestCase_SixPrismsShareingAnEdge.zip](https://github.com/user-attachments/files/29750075/1729_TestCase_SixPrismsShareingAnEdge.zip)
this is a simple 2mm³ triangular prism rotated 6 times along one of its edges. You can increase the number of repeats in the step-rotating group up to 25 - when the total volume becomes 50mm³ without anything breaking. Try the same with master before this PR or with 3.2.

## Here is a full list of everything that this fixes - I personally tested every model:

### Fixed in this PR:

- https://github.com/solvespace/solvespace/issues/683#issuecomment-683274604
  - https://github.com/solvespace/solvespace/files/5145107/wrong2.zip - works in both 3.2 (sometimes, based on stack junk?) and here.

- https://github.com/solvespace/solvespace/issues/683#issuecomment-730602943
  - https://github.com/solvespace/solvespace/files/5569373/wrong3.zip

- https://github.com/solvespace/solvespace/issues/683#issuecomment-731718242
  - https://github.com/solvespace/solvespace/files/5579199/wrong4.zip

- https://github.com/solvespace/solvespace/issues/728#issue-713235181
  - https://github.com/solvespace/solvespace/files/5315184/gusseted-truss.zip

- https://github.com/solvespace/solvespace/issues/1091#issuecomment-903661168
  - https://github.com/solvespace/solvespace/files/7030785/NURBS_ThreePrisms.zip

- https://github.com/solvespace/solvespace/issues/1091#issuecomment-945190823
  - https://github.com/solvespace/solvespace/files/7360794/corner-nurbs_slvs3x.zip

- https://github.com/solvespace/solvespace/issues/1305#issuecomment-1297213481
  - https://github.com/solvespace/solvespace/files/9902223/1305MinimalPrizmCamferNURBSBug.zip

- https://github.com/solvespace/solvespace/issues/1367#issue-1698644174
  - https://github.com/solvespace/solvespace/files/11412489/Edge_Fail.zip

- https://github.com/solvespace/solvespace/issues/1390#issue-1846087458
  - https://github.com/solvespace/solvespace/files/12318087/wedge-diff-nurbs.slvs.gz

- https://github.com/solvespace/solvespace/issues/1398#issue-1866321687
  - https://github.com/solvespace/solvespace/files/12435894/nurbs-2.slvs.gz

- https://github.com/solvespace/solvespace/pull/1410#issuecomment-1731480056
  - https://github.com/solvespace/solvespace/files/12700083/IntersectionReallyReallyReallyReallyNastyTest_DifferenceFailsBecauseItShould.zip - my  #favorite creation :-)

- https://github.com/solvespace/solvespace/issues/1619#issue-3328296115
  - https://github.com/user-attachments/files/21822164/Issue_OpenMP.zip

### Zip files with some models fixed and some not:

- https://github.com/solvespace/solvespace/issues/35#issuecomment-531602706
  - https://github.com/solvespace/solvespace/files/3614327/NURBSTests.zip
    - IntersectionReallyReallyReallyNastyTest_DifferenceFailsBecauseItShould.slvs - fails in 3.2, **fixed here**
    - IntersectionDifferenceFailsRandomly.slvs - still fails here (maybe this is "couldn't find an ear! fail"? See below)
    - all others work in both 3.2 and here.

- https://github.com/solvespace/solvespace/issues/1291#issuecomment-1288057464
  - https://github.com/solvespace/solvespace/files/9846083/CylinderTangentFromTheInside_AKA_Fillet_NURBSTests.zip
    - PrismChamferMinimalTestCase.slvs - **fixed here**
    - PrismFilletMinimalTestCase.slvs - different issue - still fails (tangent curved surfaces)
    - PrismFilletMinimalTestCase_OnlyOneFailingEdge.slvs - different issue - still fails (tangent curved surfaces)

- https://github.com/solvespace/solvespace/issues/1305#issuecomment-1294219759
  - https://github.com/solvespace/solvespace/files/9883625/Cluster_set.zip
    - cluster_repro.slvs - fails in 3.2, **fixed here**
    - Case_error_occured.slvs - works in both 3.2 and here
    - Case_solved.slvs - fails in both 3.2 and here


### Still failing - but this is a different issue ("couldn't find an ear! fail")

- https://github.com/solvespace/solvespace/issues/537#issuecomment-669068154
  - https://github.com/solvespace/solvespace/files/5027208/Prisms_EarFail.zip

- https://github.com/solvespace/solvespace/pull/594#issuecomment-625587508
  - https://github.com/solvespace/solvespace/files/4596617/CouldntFindAnEar_NURBSTests.zip

- https://github.com/solvespace/solvespace/issues/683#issue-683996514
  - https://github.com/solvespace/solvespace/files/5112393/wrong.zip

<details>
  <summary>Details</summary>

// When I first discovered the using of junk on the stack (but did not realize what it was): @ruevs ruevs on Nov 22, 2020
https://github.com/solvespace/solvespace/issues/683#issuecomment-731763446

  
// Dot products: @ruevs ruevs on Oct 25, 2022
https://github.com/solvespace/solvespace/issues/1291#issuecomment-1289998801
  
</details>

File: src/srf/surfinter.cpp [hunk 3]
```diff
@@ -44,12 +44,17 @@ void SSurface::AddExactIntersectionCurve(SBezier *sb, SSurface *srfB,
...
+                sc.pts.Add(v);
+            }
+        }
     }// end omp critical
     if(existing) {
-        SCurvePt *v;
-        for(v = existing->pts.First(); v; v = existing->pts.NextAfter(v)) {
-            sc.pts.Add(v);
-        }
         if(backwards) sc.pts.Reverse();
```
Comment (phkahler) [inline src/srf/surfinter.cpp:58]: Should the next line that may reverse be moved inside the critical section too?

if(backwards) sc.pts.Reverse();

File: src/srf/raycast.cpp (hunk 1 above)
Comment (phkahler) [reply src/srf/raycast.cpp:525]: Regions of a single shell that meet along an edge must alternate inside/outside/inside/outside  otherwise that shell isn't valid. Once we bring a second shell into the picture things get complicated since the two (assuming only one new coincident edge) new surfaces can lie anywhere relative to the originals. I'm not sure I can reason about this very well, but the comment looks correct. I'm not sure how it's finding the closest one yet...

**Comment by phkahler** [comment]
This is astonishing.  @BoykoNeov is it possible to see the conversation/prompts that lead to this? Had you studied the code and problems much prior to engaging the AI?  There are bits of spatial reasoning in there that I would not expect an LLM to understand because to me they are visual representations. Here I thought a mathematician dropped in to solve some of our bugs, but how much was you and how much was AI?

**Comment by phkahler** [comment]
BTW I like the test that uses volume rather than an image. Images all need to be updated if we change the colors for example.

**Comment by ruevs** [comment]
> This is astonishing. @BoykoNeov is it possible to see the conversation/prompts that lead to this? Had you studied the code and problems much prior to engaging the AI? There are bits of spatial reasoning in there that I would not expect an LLM to understand because to me they are visual representations. Here I thought a mathematician dropped in to solve some of our bugs, but how much was you and how much was AI?

This PR was prompted (pun intended) by a conversation about vibe coding @BoykoNeov and I had (in person) so I think I know the answer, but I'll let him answer :-)

<details>
In the meantime - I'll start planning for retirement and/or dealing with a Terminator/WALL-E/Blade Runner/The Matrix etc. scenarios ;-) 'been thinking about it for the last <a href="https://en.wikipedia.org/wiki/Attention_Is_All_You_Need">nine years</a>...
</details>

**Comment by phkahler** [comment]
Can't keep hand off that "Rebase and Merge" button.....

**Comment by phkahler** [comment]
Fixes #1398

I checked a LOT of sketches and saw no regressions.

**Comment by phkahler** [comment]
AFAICT this doesn't fix the original issue in #683 in wrong.zip.

**Comment by ruevs** [comment]
> AFAICT this doesn't fix the original issue in #683 in wrong.zip.

You are right - it does not fix it. I did not read carefully enough - one needs to delete an re-create the extrusion to reproduce the bug. I corrected my comment.

> Fixes #1398

Indeed! But **why** does it fix that?! Probably the junk on the stack. I'll have to debug...
EDIT: I debugged - it is the exact same problem. And I confirmed it by always returning a "wrong" classification in the `} else { // Our edge is tangent to the shell at the shell's edge:` - and it always fails. Good.

> Can't keep hand off that "Rebase and Merge" button.....

:-) Indeed. If/when we merge this a 3.3 release is immediately in order, in my opinion (after we fix the again [failing macOS build](https://github.com/solvespace/solvespace/actions/runs/28905261211/job/85750859358#step:3:13)).

But please wait a bit.

I want to see if @jwesthues will take a look at the "alternating knife edges" logic
image

**Comment by BoykoNeov** [comment]
> This is astonishing. @BoykoNeov is it possible to see the conversation/prompts that lead to this? Had you studied the code and problems much prior to engaging the AI? There are bits of spatial reasoning in there that I would not expect an LLM to understand because to me they are visual representations. Here I thought a mathematician dropped in to solve some of our bugs, but how much was you and how much was AI?
 
 I am a very basic programmer and here - practically a complete layman - it was 100% LLMs work. As ruevs noted - it was just born out of a conversation about vibe coding (with which I am experimenting in the past month). Practically theese are all my meaningful prompts lol :
 
solvespace/solvespace#1619 - investigate the issue, replicate if needed, propose and test solutions.
commit this on a branch and draft a PR locally for now
investigate the open ends
add the concave model as a testsuite case
why was the ruevs case not fixed?
ok , work on 1091 now

Here is the trimmed transcript of the sessions, alas - the internal thinking is encrypted and Fable 5 shares relatively few (compared to Opus) of its thaughts verbally.
https://gist.github.com/BoykoNeov/17dd078385c634db338355a75acd1050

If you want to test what Fable can do on other issues, say the word, we have until 12th of July (or when I run out of tokens :) )

**Comment by ruevs** [comment]
> https://gist.github.com/BoykoNeov/17dd078385c634db338355a75acd1050

Yep. The LLM is **_truly_** impressive.

A small example from the log:

```
**The case ruevs pointed at in #1291 — the missing `dotp[0] > TOL && dotp[1] < -TOL` branch — *is* what our fix fixes.** That comment describes exactly the tangent edge-on-edge path in `ClassifyEdge()` where the two dot products have opposite signs and the function used to fall through with `return false` and unwritten outputs. In the #1619 thread ruevs said "I fixed this case, but not [#1091] — it is harder," and then retracted: "My 'fix' is not good in the general case. More work needed." He never published his patch, so I can't say exactly what was wrong with it — but the most likely gap is the one our version handles explicitly: the classification isn't a constant in that branch. It depends on whether the shell's edge is convex or concave (reflex) — convex → both sides `SURF_OUTSIDE`, concave → both sides `SURF_INSIDE`. A fix that always picks one answer works on the common convex model (like the #1619 repro) and silently produces wrong geometry on a concave tangency — which is exactly why the new testsuite case exercises both branches at the same edge.
```

and this is exactly what I missed and exactly why my fix (that I never published) did not work in all cases...

**Comment by ruevs** [comment]
Interestingly this does **not** fix https://github.com/solvespace/solvespace/issues/1452 and https://github.com/solvespace/solvespace/issues/1401

**Comment by ruevs** [comment]
> If you want to test what Fable can do on other issues, say the word, we have until 12th of July (or when I run out of tokens :) )

@BoykoNeov If you want to burn more tokens on SolveSpace, then continuing from this branch (and context?) ask the model to work on https://github.com/solvespace/solvespace/issues/1291 next,  specifically concentrating on the `PrismFilletMinimalTestCase_OnlyOneFailingEdge.slvs` file from here https://github.com/solvespace/solvespace/files/9846083/CylinderTangentFromTheInside_AKA_Fillet_NURBSTests.zip

**Comment by phkahler** [comment]
> ask the model to work on #1291 next

@ruevs and @BoykoNeov 

That is the exact one I was going to point to.

**Comment by BoykoNeov** [comment]
@phkahler @ruevs ok, this evening I will point it to 1291 and we will see. I pointed it now to 1730 
https://github.com/solvespace/solvespace/pull/1730

**Comment by BoykoNeov** [comment]
@phkahler  @ruevs 
  https://github.com/solvespace/solvespace/pull/1731

File: src/srf/surfinter.cpp (hunk 3 above)
Comment (ruevs) [reply src/srf/surfinter.cpp:58]: As @phkahler pointed out in #1728 - no

**Comment by phkahler** [comment]
@ruevs  I understand the logic now. You provided a picture where many edges are coincident in a single shell. If you go around the edge say 1/2 radius, you'll pass through all the faces that intersect that central edge and you'll alternate inside/outside/inside... (assuming Z depth is correct) Those faces normal vector is going to point CC or CCW depending which side is inside/outside.   So if that's shell A and we want to merge it with shell B, we need to classify every face of B that shares that central edge. For any face of B we just need to find the nearest of the radial faces of A and decide if the B face is on the inside or outside of that face from A. It might also be coincident, in which case we need to know if the normal is same or opposite.

It's a very weird special case, but also very common in CAD. We can of course treat any coincident edges the same way even if shell A only has 2 faces along the edge, those from B can still be classified relative to A with this method. Ummm does that mean it replaces all the existing coincident edge classification?

**Comment by ruevs** [comment]
## This is merged - as part of merging the first nine commits in #1731, the first four of which are the ones from here.

Closing

@phkahler the ones from the eight linked issues (on the right) - that are fully fixed - will have to be closed by hand. I'm out of time for now - please do it. My comment above will be helpful - it lists which test models are fixed.
